backup: run tests with DRPC randomly enabled

suj-krishnan opened the pull request:
This change introduces a package-level setting that randomly enables DRPC in pkg/backup.

Epic: [CRDB-48935](https://cockroachlabs.atlassian.net/browse/CRDB-48935)
Fixes: https://github.com/cockroachdb/cockroach/issues/155569
Release note: None

**Notes:**
- Manual verification: Since the setting turns on/off randomly during test runs, the following command was used to manually verify that all combinations of the newly added setting and the existing  tenant setting (that is also enabled randomly) pass.
 `./dev test pkg/backup/<subdir> -v -- --test_env=COCKROACH_TEST_TENANT=<true/false/shared/external> --test_env=COCKROACH_TEST_DRPC=<true/false> 
`
- The following tests are disabled in this PR because they currently fail with the following error:
`"grpc: could not fetch metadata or no filename in metadata"`
This is a known issue that has been root-caused to differences in how metadata is handled between grpc and drpc and work to fix this issue is in progress as per @shubhamdhama. Relevant slack thread - https://cockroachlabs.slack.com/archives/C06V95X6CVA/p1761751554364579
The tests can be enabled after the issue is fixed.
Disabled tests:
TestBackupRestoreResolveDestination
All tests directly under pkg/backup

Nukitt reviewed (approved):
One minor comment about the commit message otherwise  <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @Nukitt reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157188#-:-Odqg2x-2Nn3hw7XnmA2:b-ov3rw5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @cthumuluru-crdb, @kev-cao, and @shubhamdhama)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157188#-OdqbBVK-iDiYdLAzj3K:-OdqbBVK-iDiYdLAzj3L:bfbxcam):*
Please update the commit description to have content similar to the PR description you have.  Also, in "informs"/"fixes" we tend to add the github issue since that can mark the issue closed automatically once the pr gets merged.


<!-- Sent from Reviewable.io -->

suj-krishnan committed:
backup: run tests with DRPC randomly enabled

This change introduces a package-level setting that randomly enables DRPC in pkg/backup.

Epic: CRDB-48935
Informs: https://cockroachlabs.atlassian.net/browse/CRDB-55479
Release note: None

cthumuluru-crdb commented:
> > @suj-krishnan, anything pending in #155569 after this fix besides the failing test that's disabled?
> 
> @cthumuluru-crdb Will have to re-run the disabled tests after the drpc metadata issue is fixed to see if there are any new failures in them. But otherwise, no.

In that case, maybe you can mark #155569 as fixed and open a separate ticket to re-enable skipped tests once DRPC metadata issue is fixed?

cthumuluru-crdb reviewed (approved):
lgtm!
